libvips/foreign/spngsave.c

```diff
@@ -0,0 +1,915 @@
+			vips_foreign_save_spng_pack( spng,
+				spng->line, line, sizeof_line );
+			line = spng->line;
+			sizeof_line = spng->sizeof_line;
+		}
+
+		if( (error = spng_encode_row( spng->ctx, line, sizeof_line )) )
+			break;
+	}
+
+	/* You can get SPNG_EOI for the final scanline.
+	 */
+	if( error && 
+		error != SPNG_EOI ) {
+		vips_error( class->nickname, "%s", spng_strerror( error ) ); 
+		return( -1 );
+	}
+
+	return( 0 );
+}
+
+static int
+vips_foreign_save_spng_write( VipsForeignSaveSpng *spng, VipsImage *in ) 
+{
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( spng );
+	VipsForeignSave *save = (VipsForeignSave *) spng;
+
+	int error;
+	struct spng_ihdr ihdr;
+	struct spng_phys phys;
+	struct spng_plte plte = { 0 };
+	struct spng_trns trns = { 0 };
+	int fmt;
+	enum spng_encode_flags encode_flags;
+
+	spng->ctx = spng_ctx_new( SPNG_CTX_ENCODER );
+
+	if( (error = spng_set_png_stream( spng->ctx, 
+		vips_foreign_save_spng_write_fn, spng )) ) {
+		vips_error( class->nickname, "%s", spng_strerror( error ) ); 
+		return( -1 );
+	}
+
+#ifdef HAVE_IMAGEQUANT
+	if( spng->palette ) {
+		VipsImage *im_index;
+		VipsImage *im_palette;
+		int palette_count;
+		int i;
+
+		if( vips__quantise_image( in, &im_index, &im_palette, 
+			1 << spng->bitdepth, 
+			spng->Q, 
+			spng->dither, 
+			spng->effort, 
+			FALSE ) )
+			return( -1 );
+
+		/* PNG is 8-bit index only.
+		 */
+		palette_count = im_palette->Xsize;
+		g_assert( palette_count <= 256 );
+
+		for( i = 0; i < palette_count; i++ ) {
+			VipsPel *p = (VipsPel *) 
+				VIPS_IMAGE_ADDR( im_palette, i, 0 );
+
+			if( p[3] == 255 ) {
+				struct spng_plte_entry *entry = 
+					&plte.entries[plte.n_entries];
+
+				entry->red = p[0];
+				entry->green = p[1];
+				entry->blue = p[2];
+				plte.n_entries += 1;
+			}
+			else {
+				trns.type3_alpha[trns.n_type3_entries] = p[3];
+				trns.n_type3_entries += 1;
+			}
+		}
+
+#ifdef DEBUG
+		printf( "attaching %d entry palette\n", plte.n_entries );
+		if( trns.n_type3_entries )
+			printf( "attaching %d transparency values\n", 
+			     trns.n_type3_entries );
+#endif /*DEBUG*/
+
+		VIPS_UNREF( im_palette );
+
+		in = spng->memory = im_index;
+	}
+#endif /*HAVE_IMAGEQUANT*/
+
+	ihdr.width = in->Xsize;
+	ihdr.height = in->Ysize;
+	ihdr.bit_depth = spng->bitdepth;
+
+	/* Low-bitdepth write needs an extra buffer for packing pixels.
+	 */
+	if( spng->bitdepth < 8 ) {
+		spng->sizeof_line = 1 + VIPS_IMAGE_SIZEOF_LINE( in ) / 
+			(8 / spng->bitdepth);
+
+		if( !(spng->line = 
+			vips_malloc( NULL, VIPS_IMAGE_SIZEOF_LINE( in ) )) )
+			return( -1 );
+	}
+
+	switch( in->Bands ) {
+	case 1:
+		if( spng->palette )
+			ihdr.color_type = SPNG_COLOR_TYPE_INDEXED; 
+		else
+			ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE;
+		break;
+
+	case 2:
+		ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE_ALPHA; 
+		break;
+
+	case 3:
+		ihdr.color_type = SPNG_COLOR_TYPE_TRUECOLOR;
+		break;
+
+	case 4: 
+		ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE_ALPHA; 
```

Comment:
This change resolves a lot of test failures on the [images.weserv.nl test suite](https://github.com/weserv/images/tree/5.x/test).
```suggestion
		ihdr.color_type = SPNG_COLOR_TYPE_TRUECOLOR_ALPHA; 
```

But somehow it still produces transparent images on test cases where the alpha channel is temporarily separated, for example:
```bash
$ curl -LO https://github.com/weserv/images/raw/5.x/test/api/fixtures/alpha-layer-1-fill.png

# Separate alpha channel
$ vips extract_band alpha-layer-1-fill.png rgb.v 0 --n 3
$ vips extract_band alpha-layer-1-fill.png a.v 3

# Change brightness
$ vips linear rgb.v x.v 1 76

# Bandjoin
$ vips bandjoin "x.v a.v" x.png
```

Reply:
Replacing the last command with:
```bash
# Cast float image to uchar
$ vips cast x.v x2.v uchar

# Bandjoin
$ vips bandjoin "x2.v a.v" x.png
```

Does produce a correct image. So I guess this cast is also needed in `spngsave.c`:
https://github.com/libvips/libvips/blob/de199fe59cde38290709c7eb0d25406daa1314d7/libvips/foreign/pngsave.c#L115-L129

```bash
$ vipsheader x.v
x.v: 2048x1536 float, 3 bands, srgb, pngload
```

Reply:
Oooof, that was a dumb error, good spot.

I revised the 8/16-bit save rules and the float-save case seems to work now. Thanks!